transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,598 @@
+    channel: EndpointChannel,
+    // Channel to receive messages from the transport or connections.
+    rx: mpsc::Receiver<ToEndpoint>,
+
+    // Socket for sending and receiving datagrams.
+    socket: Arc<P::Socket>,
+    // Future for writing the next packet to the socket.
+    next_packet_out: Option<(
+        Pin<Box<dyn Future<Output = Result<usize, io::Error>> + Send>>,
+        usize,
+    )>,
+    // Stream of inbound datagrams.
+    receive_stream: ReceiveStream<P>,
+
+    // List of all active connections, with a sender to notify them of events.
+    alive_connections:
+        HashMap<quinn_proto::ConnectionHandle, mpsc::Sender<quinn_proto::ConnectionEvent>>,
+    // Channel to forward new inbound connections to the transport.
+    // `None` if server capabilities are disabled, i.e. the endpoint is only used for dialing.
+    new_connection_tx: Option<mpsc::Sender<Connection>>,
+    // Whether the transport dropped its handle for this endpoint.
+    is_orphaned: bool,
+}
+
+impl<P: Provider> EndpointDriver<P> {
+    fn new(
+        endpoint_config: Arc<quinn_proto::EndpointConfig>,
+        client_config: quinn_proto::ClientConfig,
+        new_connection_tx: Option<mpsc::Sender<Connection>>,
+        server_config: Option<Arc<quinn_proto::ServerConfig>>,
+        channel: EndpointChannel,
+        socket: P::Socket,
+        rx: mpsc::Receiver<ToEndpoint>,
+    ) -> Self {
+        let socket = Arc::new(socket);
+        EndpointDriver {
+            endpoint: quinn_proto::Endpoint::new(endpoint_config, server_config),
+            client_config,
+            channel,
+            rx,
+            socket: socket.clone(),
+            next_packet_out: None,
+            receive_stream: ReceiveStream::new(socket),
+            alive_connections: HashMap::new(),
+            new_connection_tx,
+            is_orphaned: false,
+        }
+    }
+
+    /// Insert future to send a datagram on the socket.
+    fn send_packet_out(&mut self, transmit: quinn_proto::Transmit) {
+        let len = transmit.contents.len();
+        let socket = self.socket.clone();
+        let send =
+            async move { P::send_to(&socket, &transmit.contents, transmit.destination).await }
+                .boxed();
+        self.next_packet_out = Some((send, len));
+    }
+
+    /// Handle a message sent from either the [`QuicTransport`](super::QuicTransport) or a [`Connection`].
+    fn handle_message(&mut self, to_endpoint: ToEndpoint) -> ControlFlow<()> {
+        match to_endpoint {
+            ToEndpoint::Dial { addr, result } => {
+                // This `"l"` seems necessary because an empty string is an invalid domain
+                // name. While we don't use domain names, the underlying rustls library
+                // is based upon the assumption that we do.
+                let (connection_id, connection) =
+                    match self.endpoint.connect(self.client_config.clone(), addr, "l") {
```

Comment:
`client_config` should be with an empty peer ID if we are dialing an address without peer ID. But if we are dialing with `addr/p2p/ID`, a new client config with an appropriate peer ID should be created. See https://github.com/libp2p/rust-libp2p/pull/2947.

Reply:
Will add this once #2945 is merged and we can use tls from there.

Reply:
I think this PR won't be merged before #2945 is merged.